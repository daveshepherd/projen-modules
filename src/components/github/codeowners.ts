import { IConstruct } from 'constructs';
import { TextFile } from 'projen';
import { NodeProject } from 'projen/lib/javascript';

export class CodeOwners {
  constructor(project: IConstruct, codeOwners: Array<string>) {
    if (project instanceof NodeProject) {
      project.npmignore?.addPatterns('CODEOWNERS');
    }

    new TextFile(project, 'CODEOWNERS', {
      lines: codeOwners.map((owner) => `* @unibuddy/${owner}`),
    });
  }
}
