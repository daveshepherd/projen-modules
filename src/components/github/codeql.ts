import { Component, Project, YamlFile } from 'projen';

export class CodeQL extends Component {
  private yamlFile?: YamlFile;

  constructor(project: Project) {
    super(project);
    this.createYamlFile();
  }

  private createYamlFile() {
    if (this.yamlFile == null) {
      this.yamlFile = new YamlFile(
        this.project,
        '.github/workflows/codeql.yml',
        {
          obj: {
            name: 'CodeQL Advanced',
            on: {
              push: {
                branches: ['main'],
              },
              pull_request: {
                branches: ['main'],
              },
              schedule: [{ cron: '44 12 * * 1' }],
            },
            jobs: {
              analyze: {
                'name': 'Analyze (${{ matrix.language }})',
                'runs-on':
                  "${{ (matrix.language == 'swift' && 'macos-latest') || 'ubuntu-latest' }}",
                'permissions': {
                  'security-events': 'write',
                  'packages': 'read',
                  'actions': 'read',
                  'contents': 'read',
                },
                'strategy': {
                  'fail-fast': false,
                  'matrix': {
                    include: [
                      {
                        'language': 'javascript-typescript',
                        'build-mode': 'none',
                      },
                      { 'language': 'actions', 'build-mode': 'none' },
                    ],
                  },
                },
                'steps': [
                  { name: 'Checkout repository', uses: 'actions/checkout@v4' },
                  {
                    name: 'Initialize CodeQL',
                    uses: 'github/codeql-action/init@v3',
                    with: {
                      'languages': '${{ matrix.language }}',
                      'build-mode': '${{ matrix.build-mode }}',
                    },
                  },
                  {
                    name: 'Perform CodeQL Analysis',
                    uses: 'github/codeql-action/analyze@v3',
                    with: {
                      category: '/language:${{matrix.language}}',
                    },
                  },
                ],
              },
            },
          },
          committed: true,
        },
      );
    }
  }
}
