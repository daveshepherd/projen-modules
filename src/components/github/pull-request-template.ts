export const DEFAULT_PULL_REQUEST_TEMPLATE = [
  '### Background',
  '_Short description of issue being resolved or a feature being added._',
];
