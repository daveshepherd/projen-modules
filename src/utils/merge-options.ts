const isObject = (obj: unknown) => obj && typeof obj === 'object';

/**
 * Performs a deep merge of objects and returns new object. Does not modify
 * objects (immutable) and merges arrays via concatenation.
 *
 * @param  objects - Objects to merge
 * @returns  New object with merged key/values
 */
export function mergeOptions<
  Defaults extends Record<string, any>,
  Options extends Record<string, any>
>(defaults: Defaults, options?: Options): Defaults & Options {
  return [defaults, options].reduce<Record<string, any>>((prev, obj) => {
    if (obj === undefined) {
      return prev;
    }

    const result: Record<string, any> = { ...prev };
    Object.keys(obj).forEach((key) => {
      const pVal = prev[key];
      const oVal = obj[key];

      if (Array.isArray(pVal) && Array.isArray(oVal)) {
        result[key] = [...pVal, ...oVal];
      } else if (isObject(pVal) && isObject(oVal)) {
        result[key] = mergeOptions(pVal, oVal);
      } else {
        result[key] = oVal;
      }
    });

    return result;
  }, {}) as Defaults & Options;
}
