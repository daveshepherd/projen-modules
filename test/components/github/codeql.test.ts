import { synthSnapshot } from 'projen/lib/util/synth';
import { CodeQL } from '../../../src/components/github/codeql';
import { TestProject } from '../../utils/test-project';

describe('CodeQL', () => {
  it('should generate codeql.yml file', () => {
    // GIVEN
    const project = new TestProject();
    // WHEN
    new CodeQL(project);
    const snapshot = synthSnapshot(project);
    // THEN
    expect(snapshot['.github/workflows/codeql.yml']).toBeDefined();
    expect(snapshot['.github/workflows/codeql.yml']).toMatchSnapshot();
  });

  it('should contain the correct content in codeql.yml', () => {
    // GIVEN
    const project = new TestProject();
    // WHEN
    new CodeQL(project);
    const snapshot = synthSnapshot(project);
    // THEN
    const content = snapshot['.github/workflows/codeql.yml'];
    expect(content).toContain('name: CodeQL Advanced');
    expect(content).toContain('on:');
    expect(content).toContain('jobs:');
  });
});
