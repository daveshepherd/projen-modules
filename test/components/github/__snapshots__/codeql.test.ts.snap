// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeQL should generate codeql.yml file 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: CodeQL Advanced
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  schedule:
    - cron: 44 12 * * 1
jobs:
  analyze:
    name: Analyze (\${{ matrix.language }})
    runs-on: \${{ (matrix.language == 'swift' && 'macos-latest') || 'ubuntu-latest' }}
    permissions:
      security-events: write
      packages: read
      actions: read
      contents: read
    strategy:
      fail-fast: false
      matrix:
        include:
          - language: javascript-typescript
            build-mode: none
          - language: actions
            build-mode: none
    steps:
      - name: Checkout repository
        uses: actions/checkout@v4
      - name: Initialize CodeQL
        uses: github/codeql-action/init@v3
        with:
          languages: \${{ matrix.language }}
          build-mode: \${{ matrix.build-mode }}
      - name: Perform CodeQL Analysis
        uses: github/codeql-action/analyze@v3
        with:
          category: /language:\${{matrix.language}}
"
`;
